annotation-processing/src/main/java/com/exactpro/epfast/annotation/processing/FastProcessor.java

```diff
@@ -9,122 +12,306 @@
+                        fastPackageFactory.addPackageElement(packageElement);
+                    }
+                } else {
+                    String javaPackage = elementUtils.getPackageOf(element).toString();
+                    if (packagesForFastTypeElements.contains(javaPackage)) {
+                        reportDuplicates(entry.getKey(), entry.getValue());
+                    } else {
+                        packagesForFastTypeElements.add(javaPackage);
+                        noPackageFastTypes.computeIfAbsent(javaPackage, key -> new FastPackageElement(javaPackage))
+                            .addElement(entry.getKey(), element);
+                    }
+                }
+
+            }
+        }
+        String rootPackage = getRootPackage(noPackageFastTypes.values().stream()
+            .map(FastPackageElement::getName)
+            .collect(Collectors.toList()));
+        fastPackageFactory.setRootPackage(rootPackage);
+        noPackageFastTypes.putIfAbsent(rootPackage, new FastPackageElement(rootPackage));
+        for (Map.Entry<String, FastPackageElement> entry : noPackageFastTypes.entrySet()) {
+            String packageName = entry.getKey();
+            if (fastPackageFactory.getFastPackage(packageName) != null) {
+                fastPackageFactory.removePackage(packageName);
+                messager.printMessage(Diagnostic.Kind.ERROR,
+                    String.format("Java package %s matches @FastPackage name", packageName));
             } else {
-                filteredNameClassMap.put(fastTypeName, element);
+                fastPackageFactory.addPackageElement(entry.getValue());
+            }
+        }
+
+    }
+
+    String getRootPackage(Collection<String> packages) {
+        if (packages.isEmpty()) {
+            return "";
+        }
+        String root = null;
+        for (String javaPackage : packages) {
+            if (root == null) {
+                root = javaPackage;
+                continue;
             }
+            root = getDeepestCommonPackagePath(root, javaPackage);
         }
-        return filteredNameClassMap;
+        return root;
+    }
+
+    private String getDeepestCommonPackagePath(String a, String b) {
+        //stores longest common prefix
+        String[] aPackages = a.split("\\.");
+        String[] bPackages = b.split("\\.");
+        int common = -1;
+        for (int i = 0; i < Math.min(aPackages.length, bPackages.length); i++) {
+            if (!aPackages[i].equals(bPackages[i])) {
+                break;
+            }
+            common = i;
+        }
+
+        return Arrays.stream(aPackages).limit(common + 1).collect(Collectors.joining("."));
+    }
+
+    private String getFastPackageOf(Element element) {
+        PackageElement currentPackage = elementUtils.getPackageOf(element);
+        while (true) {
+            FastPackage fastPackage = currentPackage.getAnnotation(FastPackage.class);
```

Comment:
Similar code for getting fast package name is used during package names collection. Make method from it